### PR TITLE
Annotator module ts migration 2

### PR DESCRIPTION
**Pull request by acelaya**
This PR is a continuation of https://github.com/hypothesis/client/pull/5356, migrating the rest of the `annotator` module files to TypeScript.

### CODE REVIEW

File: src/annotator/frame-observer.ts [hunk 1]
```diff
@@ -47,20 +57,16 @@ export class FrameObserver {
...
-  async _addFrame(frame) {
+  private async _addFrame(frame: HTMLIFrameElement) {
     this._annotatableFrames.add(frame);
     try {
       await onNextDocumentReady(frame);
       if (this._isDisconnected) {
         return;
       }
       const frameWindow = frame.contentWindow;
-      // @ts-expect-error
```
Comment (lyzadanger) [inline src/annotator/frame-observer.ts:61]: Huh, I'm curious what the error was...was it because `frameWindow` might be undefined?

File: src/annotator/range-util.ts [hunk 2]
```diff
@@ -58,26 +51,27 @@ export function forEachNodeInRange(range, callback) {
...
  */
-export function getTextBoundingBoxes(range) {
-  const whitespaceOnly = /^\s*$/;
-  const textNodes = /** @type {Text[]} */ ([]);
+export function getTextBoundingBoxes(range: Range): DOMRect[] {
+  const textNodes: Text[] = [];
   forEachNodeInRange(range, node => {
-    if (
-      node.nodeType === Node.TEXT_NODE &&
-      !(/** @type {string} */ (node.textContent).match(whitespaceOnly))
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/annotator/range-util.ts:73]: I appreciate the encapsulation of that; see my comments about its naming and return typing...

File: src/annotator/range-util.ts [hunk 3]
```diff
@@ -132,29 +123,23 @@ export function selectionFocusRect(selection) {
...
- * @param {(n: Node) => T} itemForNode - Callback returning the item for a given node
- * @return {NonNullable<T>[]} items
+ * @param itemForNode - Callback returning the item for a given node
  */
-export function itemsForRange(range, itemForNode) {
-  /** @type {Set<Node>} */
-  const checkedNodes = new Set();
-  /** @type {Set<NonNullable<T>>} */
-  const items = new Set();
+export function itemsForRange<T>(
```
Comment (lyzadanger) [inline src/annotator/range-util.ts:128]: 😅 Oof, this is a dense little abstract function, eh?

**Review by lyzadanger — APPROVED** [review]
👍🏻

File: src/annotator/frame-observer.ts (hunk 1 above)
Comment (acelaya) [reply src/annotator/frame-observer.ts:61]: Yes, exactly.

image

I have addressed it by adding `!`, which is what we have been doing in other places.

File: src/annotator/frame-observer.ts [hunk 4]
```diff
@@ -155,17 +153,16 @@ export function onNextDocumentReady(frame) {
...
-  let pollTimer;
-  /** @type {() => void} */
-  let pollForDocumentChange;
+export function onDocumentReady(
+  frame: HTMLIFrameElement,
+  callback: (err: Error | null, document?: Document) => void,
+  { pollInterval = 10 }: { pollInterval?: number } = {}
+): () => void {
+  let pollTimer: number | undefined;
+  // eslint-disable-next-line prefer-const
```
[the remark this answers is not in the document]
Comment (acelaya) [reply src/annotator/frame-observer.ts:164]: Oh, I forgot to add a comment explaining it.

There's a conflict in linting rules. As you can see, the variable is not initialized there, so it cannot be defined as `const`.

I tried to refactor the code a bit, rearranging some of the subsequent instructions, but there are crossed dependencies between some of the callbacks following this.

I also tried initializing it with value `undefined`, but then there was other linting rules complaining.

My feeling is that the whole code block needs some refactoring, as this callback is set for the first time when another callback is executed, and then there's a third callback which relies on the second one having been executed before, which in turn means this one has a value at that point. It's hard to follow in general.

I will add a comment for now and try to address that separately.

File: src/annotator/range-util.ts [hunk 5]
```diff
@@ -58,26 +51,27 @@ export function forEachNodeInRange(range, callback) {
   }
 }
 
+function nodeIsText(node: Node): node is Text {
```
[the remark this answers is not in the document]
Comment (acelaya) [reply src/annotator/range-util.ts:54]: Actually, perhaps this should be split in two helper functions, one acting only as type guard (which is required to be able to add the object to a list of `Text` objects), and another one ensuring the content is not empty.

The second one would return just `boolean`, but only if both conditions are met, we add the object to the list.

[the remark this answers is not in the document]
Comment (acelaya) [reply src/annotator/range-util.ts:54]:
```ts
function nodeIsText(node: Node): node is Text {
  return node.nodeType === Node.TEXT_NODE;
}

function textNodeContainsText(textNode: Text): boolean {
  const whitespaceOnly = /^\s*$/;
  return !textNode.textContent!.match(whitespaceOnly);
}

[...]

if (nodeIsText(node) && textNodeContainsText(node)) {
  textNodes.push(node);
}
```

File: src/annotator/range-util.ts [hunk 6]
```diff
@@ -58,26 +51,27 @@ export function forEachNodeInRange(range, callback) {
...
-    ) {
-      textNodes.push(/** @type {Text} */ (node));
+    if (nodeIsText(node)) {
+      textNodes.push(node);
     }
   });
 
-  /** @type {DOMRect[]} */
-  let rects = [];
+  let rects: DOMRect[] = [];
```
[the remark this answers is not in the document]
Comment (acelaya) [reply src/annotator/range-util.ts:74]: Actually, it can be nicely refactored from `textNodes.forEach()` into `return textNodes.flatMap()`, with little changes, and all tests keep passing.

I have even tried to add changes just to make sure the tests failed in that case, so it looks good.

```diff
- let rects: DOMRect[] = [];
- textNodes.forEach(node => {
+ return textNodes.flatMap(node => {
  const nodeRange = node.ownerDocument.createRange();
  nodeRange.selectNodeContents(node);
  if (node === range.startContainer) {
    nodeRange.setStart(node, range.startOffset);
  }
  if (node === range.endContainer) {
    nodeRange.setEnd(node, range.endOffset);
  }
  if (nodeRange.collapsed) {
    // If the range ends at the start of this text node or starts at the end
    // of this node then do not include it.
-    return;
+    return [];
  }

  // Measure the range and translate from viewport to document coordinates
  const viewportRects = Array.from(nodeRange.getClientRects());
  nodeRange.detach();
-  rects = rects.concat(viewportRects);
+  return viewportRects;
});
- return rects;
```
